test/jdk/java/awt/regtesthelpers/PassFailJFrame.java

```diff
@@ -1393,6 +1399,77 @@ public Builder positionTestUI(PositionWindows positionWindows) {
             return this;
         }
 
+        /**
+         * Positions the test UI windows in a row to the right of
+         * the instruction frame. The top of the windows is aligned to
+         * that of the instruction frame.
```

Comment:
Since we are touching this, what are our options for positioning the instruction frame?

The `.position(PassFailJFrame.Position.TOP_LEFT_CORNER)` has no effect now with positionTestUIRightRow:

```java
        PassFailJFrame.builder()
                      .instructions(INSTRUCTIONS)
                      .rows(15)
                      .columns(30)
                      .testUI(() -> WindowCreator.createTestWindows(2))
                      .position(PassFailJFrame.Position.TOP_LEFT_CORNER)
                      .positionTestUIRightRow()
                      .build()
                      .awaitAndCheck();
```

Should we provide a way to place instruction frame in arbitrary location on screen?
It will allow us to provide more flexibility and more advanced layouts without too much effort from a test developer.

Reply:
If you call `.position(PassFailJFrame.Position.TOP_LEFT_CORNER)` after `.positionTestUIRightRow`, it is respected.

Reply:
> Should we provide a way to place instruction frame in arbitrary location on screen?

You can do it by using the `InstructionUI` interface that's passed to you when you implement a custom layout.

We're still trying to encapsulate the frame that hosts the instructions… if it's exposed, then the test developer can do whatever they want… yet then they have to ensure calling methods on that frame from EDT.

I'd rather not expose the frame at this time.

How often does one need to position the instructions at an arbitrary location on the screen? The `TOP_LEFT_CORNER` position is used by 6 tests only.

Reply:
With the instruction frame exposed, it would be possible to create owned modeless `JDialog` or `Dialog` instead of `JFrame` and `Frame` which has a great benefit: all the windows are brought to the top when you switch back to the test from another application. And there's only a single window on the taskbar for the test.

I've been thinking about adding helper methods `createOwnedJDialog` and `createOwnedDialog` to cover this use case.

Currently, it's possible to host test UI in a modeless dialog if you use `testUI` with `PanelCreator` that is pass a method which creates `JComponent`. I admit I like navigating the UI much easier in this case.

Reply:
> If you call `.position(PassFailJFrame.Position.TOP_LEFT_CORNER)` after `.positionTestUIRightRow`, it is respected.

Perhaps, this behaviour should be called out in the docs somewhere. All `.positionTestUIRight*` as well as `.positionTestUIBottom*` *implicitly* set the position.

Reply:
> You can do it by using the InstructionUI interface that's passed to you when you implement a custom layout.

But it cannot be combined with the newly added methods like:

```java
        PassFailJFrame.builder()
                .instructions(INSTRUCTIONS)
                .rows(15)
                .columns(30)
                .testUI(() -> WindowCreator.createTestWindows(3))
                .positionTestUI((testWindows, instructionUI) -> instructionUI.setLocation(200, 200))
                // or add a new method like
                //.instructionFrameLocation(200, 200)
                .positionTestUIRightRow()
                .build()
                .awaitAndCheck();
```

The idea is to make it very simple without having to implement the `positionTestUI` interface (I am not talking about exposing the frame, see the snippet above).
Like place the instruction frame here, and layout other windows relative to it.

But I agree it is a rare case.

Reply:
No, not directly.

But you can combine it with explicit call to `WindowLayouts`:

```java
.positionTestUI((testWindows, instructionUI) -> {
    instructionUI.setLocation(200, 200);
    WindowLayouts.rightOneRow(testWindows, instructionUI);
})
```

So, this (rare) scenario still doesn't require you to reimplement everything.

Reply:
Oh, I hadn't paid attention to the fact that these methods are public. I guess it is a good solution to my request.

Reply:
I like the updated version much better that what I had initially. It's really *flexible* now. One can combine different layouts on sublists to achieve more complex layouts. Yet I am still to play around with it. Anyway more complex layouts are out of scope for this PR.

Thank you @azvegint! It is your usage in #21180 that made me realise I had overcomplicated the code for laying out windows.

Reply:
Thank you for implementing this!